lib/ota/portable/nordic/nrf52840-dk/aws_ota_pal.c

```diff
@@ -515,24 +516,50 @@ static ret_code_t prvPAL_WriteImageDescriptor( OTA_FileContext_t * const C )
 /* Write a block of data to the specified file. */
 ret_code_t prvWriteFlash( uint32_t ulOffset,
                           uint32_t ulBlockSize,
-                          void * const pacData )
+                          uint8_t * const pacData )
 {
-    uint32_t ulByteIndex, ul32BitBlocksTosend, ulTotal32BitBlocks, ul32BitBlocksSent;
+    uint32_t ul32BitBlocksTosend, ulTotal32BitBlocks, ul32BitBlocksSent;
     ret_code_t xErrCode = NRF_SUCCESS;
-    ulByteIndex = 0;
-    xEventGroupClearBits( xFlashEventGrp, otapalFLASH_SUCCESS | otapalFLASH_FAILURE );
-    /* Softdevice can write only by 32-bit words */
-    xErrCode = sd_flash_write( ( uint32_t * ) ulOffset, pacData, (ulBlockSize + 3)  / 4 );
+    ulTotal32BitBlocks = (ulBlockSize+3)/4;
+    ul32BitBlocksSent = 0;
 
-    if( xErrCode == NRF_SUCCESS )
+
+    while(ul32BitBlocksSent < ulTotal32BitBlocks)
     {
-        EventBits_t xFlashResult = xEventGroupWaitBits( xFlashEventGrp, otapalFLASH_SUCCESS | otapalFLASH_FAILURE, pdTRUE, pdFALSE, portMAX_DELAY );
+      ul32BitBlocksTosend = 0;
+
+      ul32BitBlocksTosend = ulTotal32BitBlocks - ul32BitBlocksSent;
+
+      if(ul32BitBlocksTosend > otapalSERIALIZING_ARRAY_SIZE)
+      {
+        ul32BitBlocksTosend = otapalSERIALIZING_ARRAY_SIZE;
+      }
+
+      if(ul32BitBlocksTosend > NRF_FICR->CODEPAGESIZE)
+      {
+        ul32BitBlocksTosend = NRF_FICR->CODEPAGESIZE;
+      }
+
+      memcpy(pulSerializingArray, pacData, ulBlockSize); 
```

Comment:
Should this be done outside while loop ?

Reply:
Yes on those there is not choice. We can only copy as much as a page size or what is available in the buffer. The chunk is then copied in flash. The next chunk is copied in the next loop cycle